src/state/editor.rs

```diff
@@ -12,18 +13,146 @@ use orbclient::Window;
+/// A SplitBuffer and related state
+pub struct BufferInfo {
+    /// The document
+    pub raw_buffer: SplitBuffer,
     /// The current cursor
     pub current_cursor: u8,
     /// The cursors
     pub cursors: Vec<Cursor>,
-    /// The buffer (document)
-    pub buffer: SplitBuffer,
     /// The x coordinate of the scroll
     pub scroll_x: usize,
     /// The y coordinate of the scroll
     pub scroll_y: usize,
+    /// The title of the document
+    pub title: Option<String>,
+    /// True if the buffer is transient and should be deleted when
+    /// it is no longer the current buffer.
+    pub is_transient: bool,
+}
+
+impl BufferInfo {
+    /// Create a new BufferInfo with default values.
+    fn new() -> BufferInfo {
+        BufferInfo {
+            raw_buffer: SplitBuffer::new(),
+            current_cursor: 0,
+            cursors: vec![Cursor::new()],
+            scroll_x: 0,
+            scroll_y: 0,
+            title: None,
+            is_transient: false,
+        }
+    }
+}
+
+impl From<SplitBuffer> for BufferInfo {
+    fn from(b: SplitBuffer) -> BufferInfo {
+        let mut info = BufferInfo::new();
+        info.raw_buffer = b;
+
+        info
+    }
+}
+
+/// Provides access to buffer manipulation functions.
+pub struct BufferManager {
+    buffers: Vec<BufferInfo>,
+    current_buffer_index: usize,
+}
+
+impl BufferManager {
+    /// Create a new BufferManager with default values.
+    pub fn new() -> BufferManager {
+        BufferManager {
+            buffers: vec![BufferInfo::new()],
+            current_buffer_index: 0,
+        }
+    }
+
+    /// Adds the specified buffer to the set of buffers and returns
+    /// its index.
+    pub fn new_buffer(&mut self, buffer: BufferInfo) -> usize {
+        self.buffers.push(buffer);
+
+        self.buffers.len() -1
+    }
+
+    /// Returns an iterator over the buffers.
+    pub fn iter(&self) -> Iter<BufferInfo> {
+        self.buffers.iter()
+    }
+
+    /// Gets the number of buffers.
+    pub fn len(&self) -> usize {
+        self.buffers.len()
+    }
+
+    /// Get a reference to the currently open buffer.
+    pub fn current_buffer(&self) -> &SplitBuffer {
+        &self.current_buffer_info().raw_buffer
+    }
+
+    /// Get a mutable reference to the currently open buffer.
+    pub fn current_buffer_mut(&mut self) -> &mut SplitBuffer {
+        &mut self.current_buffer_info_mut().raw_buffer
+    }
+
+    /// Get a reference to the currently open buffer information.
+    pub fn current_buffer_info(&self) -> &BufferInfo {
+        &self.buffers[self.current_buffer_index]
+    }
+
+    /// Get a mutable reference to the currently open buffer information.
+    pub fn current_buffer_info_mut(&mut self) -> &mut BufferInfo {
+        &mut self.buffers[self.current_buffer_index]
+    }
+
+    /// Switch the current buffer to the specified buffer
+    pub fn switch_to(&mut self, n: usize) {
+        assert!(n < self.buffers.len(), "Buffer index out of bounds");
```

Comment:
Since this function doesn't index into `self.buffers`, you won't get a panic immediately if you pass in a bad index. I thought it might be helpful it panic'd immediately so that the caller was in the backtrace.